core/CronArchive.php

```diff
@@ -946,6 +956,21 @@ protected function processArchiveDays($idSite, $lastTimestampWebsiteProcessedDay
         return $dayArchiveWasSuccessful;
     }
 
+    private function isThereAValidArchiveForPeriod($idSite, $period, $segment = '')
```

Comment:
@diosmosis wondering: Wouldn't we need to take into consideration the date as well we want to archive? Only realising this now... like what if a user is tracking visits a few days back and this archive was invalidated but not today... then we would need to detect this right? But then I'm wondering if things get in the end slower than we benefit from it. @mattab 

There are a few complexities:
* Date might be `last2` or `last52` for example... if it was say last52 does this maybe mean we need to iterate over 52 dates/archives and check if any of them is invalidated meaning doing below logic max 52 times? (also adding lots of IO)
* If date is `last2`, but there are reports to be invalidated in the past in `invalidateArchivedReportsForSitesThatNeedToBeArchivedAgain()`, does this mean we need to dynamically need to change the date that will be used in the query to say `lastXYZ` to make sure the invalidated reports will be rearchived? 

The original purpose was to make archiving faster when you have say 1000 sites and only say 200 of those sites are tracking regularly. Meaning for 800 sites we could mostly skip launching the archiver at all. But now wondering if in the end we make all other cases slower? In general it seems like the archiver should maybe behave like this though. Goes maybe also a bit in the direction of https://github.com/matomo-org/matomo/issues/15117

Reply:
FYI also created https://github.com/matomo-org/matomo/issues/15127

Reply:
I think we could potentially use multiple dates w/ one query (`SELECT COUNT(*) WHERE date1 >= X AND date2 <= Y and period = Z AND ... == $rangePperiod->getSubperiodCount()`). Think that would work?

Reply:
I reckon we'd still need to go over several tables though and we'd need to know which data is stored where etc. So we could probably do it in a few queries for example.

Reply:
Just noticed another issue, this doesn't affect the original case where 800 sites w/o visits shouldn't have archiving launch, since there won't be any archives for those sites even though they have no visits. I guess we should also check for a single visit in a period for a site? Maybe only for the non-segment requests, then if there are visits for those we'd know there could be visits for segments.

Reply:
@tsteur  for the no visits optimization I wonder if we can just change `hadWebsiteTrafficSinceMidnightInTimezone()` to check whether there has been a visit for the site since the last time archiving was run for the site? We would save the archive start time in a new option like `CronArchive.startArchiveSite.$idSite` in `archiveSingleSite()`.

Reply:
@diosmosis not 100% sure. We would probably need to fix first https://github.com/matomo-org/matomo/issues/15127

You mean then add an empty archive when no visits for today so far basically?

Reply:
I see, yes, I guess it can't work w/o #15127. I guess we can add and look for a `'start'` flag that's inserted when an archive starts archiving (in ArchiveWriter), as was done in a previous version of this PR then check for a single visit since then.

Reply:
Actually I think we don't have to worry about specifically looking for no visits... if a site has a visit the archive will be invalidated, otherwise there might be an archiving launch for the first time in a day, but after that there will be an archive... Ok, think this is ready for another review.